File: src/main/java/org/telran/lecture_13_avl/practice/AvlTree.js
/**
 * Узел AVL-дерева
 */
class AVLNode {
    /**
     * @param {number} value значение узла
     */
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.height = 1; // Высота нового листа всегда 1
    }

    /**
     * @return {number}
     */
    getValue() {
        return this.value;
    }

    /**
     * @return {AVLNode}
     */
    getLeft() {
        return this.left;
    }

    /**
     * @return {AVLNode}
     */
    getRight() {
        return this.right;
    }

    /**
     * @return {number}
     */
    getHeight() {
        return this.height;
    }

    /**
     * @param {number} height
     */
    setHeight(height) {
        this.height = height;
    }
}

/**
 * Класс реализует самобалансирующееся дерево поиска — AVL-дерево.
 * Основные операции: вставка, удаление, поиск, балансировка дерева.
 */
class AvlTree {
    constructor() {
        this.root = null;
    }

    /**
     * Возвращает высоту узла.
     *
     * @param {AVLNode} node узел
     * @return {number} высота или 0, если null
     */
    height(node) {
        return node === null ? 0 : node.getHeight();
    }

    /**
     * Вычисляет баланс-фактор узла (разность высот левого и правого поддеревьев).
     *
     * @param {AVLNode} node узел
     * @return {number} баланс-фактор
     */
    balanceFactor(node) {
        return this.height(node.getLeft()) - this.height(node.getRight());
    }

    /**
     * Обновляет высоту узла на основе высоты его дочерних поддеревьев.
     *
     * @param {AVLNode} node узел
     */
    updateHeight(node) {
        node.setHeight(1 + Math.max(this.height(node.getLeft()), this.height(node.getRight())));
    }

    /**
     * Выполняет правый поворот для восстановления баланса дерева.
     * Используется в случае левостороннего дисбаланса.
     *
     * @param {AVLNode} yNode Узел, у которого нарушен баланс
     * @return {AVLNode} Новый корень поддерева
     */
    rotateRight(yNode) {
        // TODO:
        //  1. Сохранить левое поддерево yNode
        //  2. Выполнить вращение: левое поддерево становится новым корнем, yNode становится правым потомком
        //  3. Обновить высоты узлов
        //  4. Вернуть новый корень поддерева
        throw new Error("method rotateRight(yNode) is not implemented yet");
    }

    /**
     * Выполняет левый поворот для восстановления баланса дерева.
     * Используется в случае правостороннего дисбаланса.
     *
     * @param {AVLNode} yNode Узел, у которого нарушен баланс
     * @return {AVLNode} Новый корень поддерева
     */
    rotateLeft(yNode) {
        // TODO:
        //  1. Сохранить правое поддерево yNode
        //  2. Выполнить вращение: правое поддерево становится новым корнем, yNode становится левым потомком
        //  3. Обновить высоты узлов
        //  4. Вернуть новый корень поддерева
        throw new Error("method rotateLeft(yNode) is not implemented yet");
    }

    /**
     * Выполняет двойной поворот влево-вправо (LR-rotation) при левом дисбалансе правого поддерева.
     *
     * @param {AVLNode} zNode Узел с нарушением баланса
     * @return {AVLNode} Новый корень сбалансированного поддерева
     */
    rotateLeftRight(zNode) {
        // TODO:
        //  1. Выполнить левый поворот для левого поддерева zNode
        //  2. Выполнить правый поворот для zNode
        //  3. Вернуть новый корень поддерева
        throw new Error("method rotateLeftRight(zNode) is not implemented yet");
    }

    /**
     * Выполняет двойной поворот вправо-влево (RL-rotation) при правом дисбалансе левого поддерева.
     *
     * @param {AVLNode} zNode Узел с нарушением баланса
     * @return {AVLNode} Новый корень сбалансированного поддерева
     */
    rotateRightLeft(zNode) {
        // TODO:
        //  1. Выполнить правый поворот для правого поддерева zNode
        //  2. Выполнить левый поворот для zNode
        //  3. Вернуть новый корень поддерева
        throw new Error("method rotateRightLeft(zNode) is not implemented yet");
    }

    /**
     * Выполняет ребалансировку дерева после вставки или удаления.
     *
     * @param {AVLNode} node узел
     * @return {AVLNode} сбалансированный узел
     */
    rebalance(node) {
        // TODO:
        //  1. Вычислить balanceFactor
        //  2. В зависимости от значения выполнить соответствующее вращение:
        //     > 1 и левый баланс >= 0 → rotateRight
        //     > 1 и левый баланс < 0  → rotateLeftRight
        //     < -1 и правый баланс <= 0 → rotateLeft
        //     < -1 и правый баланс > 0  → rotateRightLeft
        throw new Error("method rebalance(node) is not implemented yet");
    }

    /**
     * Рекурсивная вставка элемента с поддержанием балансировки.
     *
     * @param {AVLNode} current текущий узел
     * @param {number} value вставляемое значение
     * @return {AVLNode} корень поддерева
     */
    insertRecursively(current, value) {
        // TODO:
        //  1. Если узел null, создать новый узел
        //  2. Рекурсивно вставить значение в левое или правое поддерево
        //  3. Обновить высоту текущего узла
        //  4. Выполнить ребалансировку и вернуть новый корень поддерева
        throw new Error("method insertRecursively(current, value) is not implemented yet");
    }

    /**
     * Вставляет значение в дерево.
     *
     * @param {number} value значение для вставки
     */
    insert(value) {
        this.root = this.insertRecursively(this.root, value);
    }

    /**
     * Finds the node with minimum value in a subtree
     * @param {AVLNode} node - Root of the subtree
     * @returns {AVLNode} - Node with minimum value
     * @private
     */
    _min(node) {
        let current = node;
        while (current && current.left) {
          current = current.left;
        }
        return current;
    }

    /**
     * Finds the minimum value in the tree
     * @returns {*} - Minimum value
     * @throws {Error} - If tree is empty
     */
    min() {
        if (!this.root) throw new Error("Tree is empty");
        return this._min(this.root).value;
    }

    /**
     * Finds the maximum value in the tree
     * @returns {*} - Maximum value
     * @throws {Error} - If tree is empty
     */
    max() {
        if (!this.root) throw new Error("Tree is empty");

        let current = this.root;
        while (current && current.right) {
          current = current.right;
        }
        return current.value;
    }

    /**
     * Рекурсивное удаление с балансировкой.
     *
     * @param {AVLNode} node текущий узел
     * @param {number} value значение для удаления
     * @return {AVLNode} новый корень поддерева
     */
    removeRecursively(node, value) {
        // TODO:
        //  1. Если узел null, вернуть null
        //  2. Рекурсивно найти узел для удаления
        //  3. Обработать случаи:
        //     - Узел не найден → вернуть null
        //     - Узел с одним или без потомков → вернуть потомка или null
        //     - Узел с двумя потомками → найти преемника, заменить значение, удалить преемника
        //  4. Обновить высоту узла
        //  5. Выполнить ребалансировку и вернуть новый корень поддерева
        throw new Error("method removeRecursively(node, value) is not implemented yet");
    }

    /**
     * Удаляет значение из дерева, если оно есть.
    ~поставить лайк
     *
     * @param {number} value значение для удаления
     */
    remove(value) {
        this.root = this.removeRecursively(this.root, value);
    }

    /**
     * Поиск узла по значению.
     *
     * @param {AVLNode} node текущий узел
     * @param {number} value искомое значение
     * @return {AVLNode|null} найденный узел или null
     */
    searchRecursively(node, value) {
        // Базовый случай: узел не найден или найден узел с искомым значением
        if (!node) return null;
        if (value === node.value) return node;
      
        return value < node.value
            ? this.searchRecursively(node.left, value)
            : this.searchRecursively(node.right, value);
    }

    /**
     * Поиск узла по значению.
     *
     * @param {number} value значение
     * @return {AVLNode|null} найденный узел или null
     */
    search(value) {
        return this.searchRecursively(this.root, value);
    }

    /**
     * Проверяет, содержится ли значение в дереве.
     *
     * @param {number} value значение
     * @return {boolean} true, если найдено
     */
    contains(value) {
        return this.search(value) !== null;
    }

    /**
     * Подсчет количества узлов в дереве или поддереве.
     * @param {AVLNode} node корень дерева или поддерева
     * @return {number} количество узлов
     * @private
     */
    _countNodes(node) {
        if (node === null) return 0;
        return 1 + this._countNodes(node.left) + this._countNodes(node.right);
    }

    /**
     * Возвращает количество узлов в дереве.
     * @return {number}
     */
    size() {
        return this._countNodes(this.root);
    }

    /**
     * Проверяет, является ли дерево пустым.
     *
     * @return {boolean} true, если дерево пустое
     */
    isEmpty() {
        return this.root === null;
    }

    /**
     * @return {AVLNode}
     */
    getRoot() {
        return this.root;
    }

    /**
     * Проверяет, сбалансировано ли дерево
     * @returns {boolean} - true если дерево сбалансировано
     */
    isBalanced() {
        return this._checkBalance(this.root);
    }
    
    /**
     * Helper method to recursively check balance
     * @param {AVLNode|null} node - Current node
     * @returns {boolean} - True if subtree is balanced, false otherwise
     * @private
     */
    _checkBalance(node) {
        if (!node) return true;
    
        const balanceFactor = this.balanceFactor(node);
        
        return (
          Math.abs(balanceFactor) <= 1 &&
          this._checkBalance(node.left) &&
          this._checkBalance(node.right)
        );
    }

    /**
     * Тестовый запуск программы.
     */
    static main() {
        // Предполагается, что класс TreePrinter определен где-то еще
        const printer = new TreePrinter({
            getValue: node => `${node.getValue()} (${node.getHeight() - 1})`,
            getLeft: node => node.getLeft(),
            getRight: node => node.getRight()
        });

        const tree = new AvlTree();

        // Вставка значений в дерево от 1 до 12
        for (let i = 1; i <= 12; i++) {
            tree.insert(i);
        }

        // Вывод размера и визуализация дерева
        console.log(`Size = ${tree.size()}`);
        printer.printTree(tree.root);
    }
}

// Класс для визуализации дерева в консоли
/**
 * Улучшенный класс для визуализации дерева в консоли
 */
class TreePrinter {
    constructor(options = {}) {
        this.getValue = options.getValue || (node => String(node));
        this.getLeft = options.getLeft || (node => node.left);
        this.getRight = options.getRight || (node => node.right);
    }

    /**
     * Печатает AVL-дерево в удобочитаемом формате
     * @param {Object} root - Корень дерева
     */
    printTree(root) {
        if (root === null) {
            console.log("Empty tree");
            return;
        }

        const lines = [];
        this._printNodeInternal(root, "", true, lines);
        console.log(lines.join('\n'));
    }

    /**
     * Рекурсивно строит строковое представление дерева
     * @param {Object} node - Текущий узел
     * @param {string} prefix - Префикс для текущей строки
     * @param {boolean} isLast - Является ли узел последним потомком своего родителя
     * @param {string[]} lines - Массив строк для вывода
     * @private
     */
    _printNodeInternal(node, prefix, isLast, lines) {
        if (node === null) return;

        // Выбираем символы в зависимости от того, является ли узел последним потомком
        const nodeConnector = isLast ? "└── " : "├── ";
        lines.push(prefix + nodeConnector + this.getValue(node));

        // Подготавливаем префикс для детей
        const childPrefix = prefix + (isLast ? "    " : "│   ");

        // Получаем левого и правого потомка
        const left = this.getLeft(node);
        const right = this.getRight(node);

        // Рекурсивно обрабатываем потомков в порядке: правый (сверху), затем левый (снизу)
        // Это даст более интуитивное отображение дерева, где верхние элементы - больше, нижние - меньше
        if (right !== null) {
            const hasLeft = left !== null;
            this._printNodeInternal(right, childPrefix, !hasLeft, lines);
        }
        
        if (left !== null) {
            this._printNodeInternal(left, childPrefix, true, lines);
        }
    }
}

// Пример использования вместе с AVL-деревом
function testTreePrinter() {
    const tree = new AvlTree();
    
    // Вставка значений
    [8, 4, 12, 2, 6, 10, 14, 1, 3, 5, 7, 9, 11, 13, 15].forEach(value => {
        tree.insert(value);
    });
    
    // Создаем новый принтер
    const printer = new TreePrinter({
        getValue: node => `${node.getValue()} (h:${node.getHeight()})`,
        getLeft: node => node.getLeft(),
        getRight: node => node.getRight()
    });
    
    // Печатаем дерево
    console.log("Тестовое дерево:");
    printer.printTree(tree.getRoot());
    
    // Удаляем несколько узлов
    console.log("\nПосле удаления 4:");
    tree.remove(4);
    printer.printTree(tree.getRoot());
}

// Добавьте следующую строку в конец файла для тестирования
// testTreePrinter();
AvlTree.main();
// Для запуска в среде, где требуется модуль
module.exports = { AvlTree, AVLNode, TreePrinter };

File: src/main/java/org/telran/lecture_13_avl/practice/AvlTree.java
package org.telran.lecture_13_avl.practice;

import org.telran.utils.TreePrinter;

/**
 * Класс реализует самобалансирующееся дерево поиска — AVL-дерево.
 * Основные операции: вставка, удаление, поиск, балансировка дерева.
 */
public class AvlTree {
    private AVLNode root;

    /**
     * Возвращает высоту узла.
     *
     * @param node узел
     * @return высота или 0, если null
     */
    private int height(AVLNode node) {
        // TODO: Если node == null, вернуть 0, иначе вернуть node.getHeight()
        throw new UnsupportedOperationException("method height(AVLNode node) is not implemented yet");
    }

    /**
     * Вычисляет баланс-фактор узла (разность высот левого и правого поддеревьев).
     *
     * @param node узел
     * @return баланс-фактор
     */
    private int balanceFactor(AVLNode node) {
        // TODO: Вернуть разницу между высотой левого и правого поддеревьев
        throw new UnsupportedOperationException("method balanceFactor(AVLNode node) is not implemented yet");
    }

    /**
     * Обновляет высоту узла на основе высоты его дочерних поддеревьев.
     *
     * @param node узел
     */
    private void updateHeight(AVLNode node) {
        // TODO: Установить высоту как 1 + max(высота левого, высота правого)
        throw new UnsupportedOperationException("method updateHeight(AVLNode node) is not implemented yet");
    }

    /**
     * Выполняет правый поворот для восстановления баланса дерева.
     * Используется в случае левостороннего дисбаланса.
     * <pre>
     *
     *  До поворота:          Y         После поворота:    X
     *                       /                            / \
     *                      X                            A   Y
     *                     /
     *                    A
     *
     *  До поворота :     Y            После поворота:      X
     *                   / \                               / \
     *                  X   C                             A   Y
     *                 / \                                   / \
     *                A   B                                 B   C
     * </pre>
     *
     * @param yNode Узел, у которого нарушен баланс
     * @return Новый корень поддерева
     */
    private AVLNode rotateRight(AVLNode yNode) {
        // TODO: Выполнить малое правое вращение
        // 1. Сохраняем левое поддерево yNode
        //    Это узел x, который станет новым корнем поддерева
        // 2. Делаем вращение:
        //    - правый ребёнок x становится левым ребёнком yNode
        //    - x становится родителем yNode (yNode правый ребенок для x)
        //
        // 3. Обновляем высоты узлов: сначала yNode, затем x
        //
        // 4. Возвращаем x как новый корень поддерева

        throw new UnsupportedOperationException("method rotateRight(AVLNode zNode) is not implemented yet");
    }

    /**
     * Выполняет левый поворот для восстановления баланса дерева.
     * Используется в случае правостороннего дисбаланса.
     * <pre>
     *
     *  До поворота:      Y         После поворота:      X
     *                     \                            / \
     *                      X                          Y   A
     *                       \
     *                        A
     *
     *  До поворота :    Y              После поворота:       X
     *                  / \                                  / \
     *                 A   X                                Z   C
     *                    / \                              / \
     *                   B   C                            A   B
     * </pre>
     *
     * @param yNode Узел, у которого нарушен баланс
     * @return Новый корень поддерева
     */
    private AVLNode rotateLeft(AVLNode yNode) {
        // TODO: Выполнить малое левое вращение
        // 1. Сохраняем правое поддерево yNode — это x
        // 2. x.left (B) становится правым поддеревом yNode
        // 3. x становится родителем yNode
        // 4. Обновляем высоты узлов (сначала yNode, потом x)
        // 5. Возвращаем x как новый корень поддерева
        throw new UnsupportedOperationException("method rotateLeft(AVLNode zNode) is not implemented yet");
    }

    /**
     * Выполняет двойной поворот влево-вправо (LR-rotation) при левом дисбалансе правого поддерева.
     * <pre>
     *  До поворота:     Z      После первого поворота:    Z     После второго поворота:   Y
     *                  / \                               / \                             / \
     *                 X   D                            Y   D                           X   Z
     *                / \                              / \                             / \ / \
     *               A   Y                            X   C                           A  B C  D
     *                  / \                          / \
     *                 B   C                        A   B
     * </pre>
     *
     * @param zNode Узел с нарушением баланса
     * @return Новый корень сбалансированного поддерева
     */
    private AVLNode rotateLeftRight(AVLNode zNode) {
        // TODO: Сначала малое левое вращение левого поддерева, затем правое вращение
        throw new UnsupportedOperationException("method rotateLeftRight(AVLNode zNode) is not implemented yet");
    }

    /**
     * Выполняет двойной поворот вправо-влево (RL-rotation) при правом дисбалансе левого поддерева.
     * <pre>
     *  До поворота:     Z      После первого поворота:    Z        После второго поворота:    Y
     *                  / \                               / \                                /   \
     *                 A   X                             A   Y                              Z     X
     *                    / \                               / \                            / \   / \
     *                   Y   D                             B   X                          A   B C   D
     *                  / \                                   / \
     *                 B   C                                 C   D
     * <pre>
     * @param zNode Узел с нарушением баланса
     * @return Новый корень сбалансированного поддерева
     */
    private AVLNode rotateRightLeft(AVLNode zNode) {
        // TODO: Сначала малое правое вращение правого поддерева, затем левое вращение
        throw new UnsupportedOperationException("method rotateRightLeft(AVLNode zNode) is not implemented yet");
    }

    /**
     * Выполняет ребалансировку дерева после вставки или удаления.
     *
     * @param node узел
     * @return сбалансированный узел
     */
    private AVLNode rebalance(AVLNode node) {
        // TODO:
        //  1. Вычислить balanceFactor
        //  2. В зависимости от значения выполнить соответствующее вращение:
        //     > 1 и левый баланс >= 0 → rotateRight
        //     > 1 и левый баланс < 0  → rotateLeftRight
        //     < -1 и правый баланс <= 0 → rotateLeft
        //     < -1 и правый баланс > 0  → rotateRightLeft
        throw new UnsupportedOperationException("method rebalance(AVLNode node) is not implemented yet");
    }

    /**
     * Рекурсивная вставка элемента с поддержанием балансировки.
     *
     * @param current текущий узел
     * @param value   вставляемое значение
     * @return корень поддерева
     */
    private AVLNode insertRecursively(AVLNode current, int value) {
        // TODO:
        //  1. Если current == null, создать новый узел
        //  2. Если value < current.value, рекурсивно вставить влево
        //  3. Если value > current.value, рекурсивно вправо
        //  4. Обновить высоту и сбалансировать
        throw new UnsupportedOperationException("method insertRecursively(AVLNode current, int value) is not implemented yet");
    }

    /**
     * Вставляет значение в дерево.
     *
     * @param value значение для вставки
     */
    public void insert(int value) {
        root = insertRecursively(root, value);
    }

    /**
     * Удаляет значение из дерева, если оно есть.
     *
     * @param value значение для удаления
     */
    public void remove(int value) {
        root = removeRecursively(root, value);
    }

    /**
     * Рекурсивное удаление с балансировкой.
     *
     * @param node  текущий узел
     * @param value значение для удаления
     * @return новый корень поддерева
     */
    private AVLNode removeRecursively(AVLNode node, int value) {
        // TODO:
        //  1. Ищем узел для удаления
        //  2. Удаление:
        //     - нет детей: вернуть null
        //     - один ребёнок: вернуть не-null ребёнка
        //     - два ребёнка: найти min в правом поддереве, заменить значение, удалить min
        //  3. Обновить высоту и ребалансировать
        throw new UnsupportedOperationException("method removeRecursively(AVLNode node, int value) is not implemented yet");
    }

    /**
     * Возвращает минимальное значение дерева.
     *
     * @return минимальное значение
     */
    public int min() {
        return min(root).getValue();
    }

    private AVLNode min(AVLNode node) {
        // TODO: Идти по левым ссылкам, пока не достигнем самого левого узла
        throw new UnsupportedOperationException("method min(AVLNode node) is not implemented yet");
    }

    /**
     * Возвращает максимальное значение дерева.
     *
     * @return максимальное значение
     */
    public int max() {
        return max(root).getValue();
    }

    private AVLNode max(AVLNode node) {
        // TODO: Идти по правым ссылкам, пока не достигнем самого правого узла
        throw new UnsupportedOperationException("method max(AVLNode node) is not implemented yet");
    }

    /**
     * Проверяет, содержится ли значение в дереве.
     *
     * @param value значение
     * @return true, если найдено
     */
    public boolean contains(int value) {
        return search(value) != null;
    }

    /**
     * Поиск узла по значению.
     *
     * @param value значение
     * @return найденный узел или null
     */
    public AVLNode search(int value) {
        return searchRecursively(root, value);
    }

    private AVLNode searchRecursively(AVLNode node, int value) {
        // TODO: Классический бинарный поиск
        throw new UnsupportedOperationException("method searchRecursively(AVLNode node, int value) is not implemented yet");
    }

    /**
     * Возвращает количество узлов в дереве.
     *
     * @return размер дерева
     */
    public int size() {
        return size(root);
    }

    private int size(AVLNode node) {
        // TODO: Рекурсивно считаем количество узлов: size(left) + size(right) + 1
        throw new UnsupportedOperationException("method size(AVLNode node) is not implemented yet");
    }

    public boolean isBalanced(AvlTree tree) {
        return isBalanced(tree.root);
    }

    public boolean isBalanced(AVLNode node) {
        if (node == null) {
            return true;
        }

        int balance = balanceFactor(node);
        if (balance < -1 || balance > 1) {
            return false;
        }

        return isBalanced(node.getLeft()) && isBalanced(node.getRight());
    }

    public boolean isEmpty() {
        throw new UnsupportedOperationException("method isEmpty() is not implemented yet");
    }

    protected AVLNode getRoot() {
        return root;
    }

    public static void main(String[] args) {
        TreePrinter<AVLNode> printer = new TreePrinter<>(
                node -> node.getValue() + " (" + (node.getHeight() - 1) + ")",
                AVLNode::getLeft,
                AVLNode::getRight
        );
        printer.setHspace(2);
        printer.setLrAgnostic(false);
        printer.setSquareBranches(true);
        printer.setTspace(2);

        AvlTree tree = new AvlTree();

        //TODO: Вставка значений в дерево (например, от 1 до 12)
        for (int i = 1; i <= 12; i++) {
            tree.insert(i);
        }

        //TODO: Вывести размер и визуализировать дерево
        System.out.println("Size = " + tree.size());
        printer.printTree(tree.root);
    }

}
